### ISSUE TITLE
custom user agent string for marketplace downloads

### ISSUE THREAD

**Issue by simonhaenisch** [Microsoft/vscode#6480]
- VSCode Version: 1.1.1
- OS Version: Windows 7

Hi, in my work environment only certain user agents are allowed to make http(s) requests, so I can't download any extension (`getaddrinfo ENOTFOUND marketplace.visualstudio.com marketplace.visualstudio.com:443`).

I had the same issue with package control for Sublime Text and there you can set a custom user agent string (https://packagecontrol.io/docs/settings) for the downloader.

Can you please implement that for VS Code?

Related: #262

**Comment by joaomoreno** [comment]
Good point. This might be related to other people's problems. Let's do it!

**Comment by joaomoreno** [comment]
Actually, I'm not so sure the problem is the `User-Agent` header...

Your machine cannot figure out to which IP address belongs `marketplace.visualstudio.com`, which should be resolved by DNS, which has nothing to do with HTTP and headers.

Check in your work environment if you need to whitelist some domains. Namely these:

```
"extensionsGallery": {
    "serviceUrl": "https://marketplace.visualstudio.com/_apis/public/gallery",
    "cacheUrl": "https://vscode.blob.core.windows.net/gallery/index",
    "itemUrl": "https://marketplace.visualstudio.com/items"
}
```

**Comment by joaomoreno** [comment]
Duplicate of #5816

**Comment by simonhaenisch** [comment]
I can access all of these domains with Firefox, because the Firefox user agent is whitelisted in my work enviroment (they even customized the user agent string for Firefox, so a default installation wouldn't work).

I tried a npm install with a custom user agent (`npm config set user_agent <string>`) and got the same `getaddrinfo ENOTFOUND registry.npmjs.org registry.npmjs.org:443` (can access the url in Firefox).

So I logged my network traffic while triggering an install in Sublime Text. The network interface receives
`HTTP CONNECT packagecontrol.io:443 HTTP/1.0`
and the answer from the proxy server is
`HTTP/1.1 407 Proxy Authentication Required (text/html)`.
With the default user agent it is followed by
`HTTP/1.1 403 Forbidden (text/html)`
but with the custom user agent there is
`HTTP/1.1 200 Connection Established`.

So this means that the custom user agent really does the trick for me there.

Then I also logged a `npm install` and I noticed that there is no direct HTTP CONNECT request to the `registry.npmjs.org` domain, but instead there is a HTTP GET request to a local `secars.dll`. After some searching it seems that in my work environment each connection that comes from `cmd.exe` goes through the Symantec Endpoint Protection firewall. It seems that no connection is allowed in the firewall settings, because the request doesn't get answered. I can't even ping any domain/ip address using `cmd.exe`. Also tried `wget` and that doesn't work either. I guess it is the same problem with VS Code.

I don't have any insight on how Node.js does HTTP requests via the network interface, but it seems to be done differently (triggered via `cmd.exe`?) than in Sublime Text/Package Control.

**Comment by joaomoreno** [comment]
Got it! Thanks for the **very** detailed explanation. Apologies for not believing in you at first.

**Comment by simonhaenisch** [comment]
@joaomoreno did you already put any work into that?

In the meantime I found out that the reason Sublime Text Package Control works with the custom user-agent string is that it uses the *Windows WinINet DLL* to make HTTP(S) requests.

see [wininet_downloader.py#L32:L33](https://github.com/wbond/package_control/blob/master/package_control/downloaders/wininet_downloader.py#L32:L33):

> A downloader that uses the Windows WinINet DLL to perform downloads. This has the benefit of utilizing system-level proxy configuration and CA certs.

It's possible to switch to different downloaders (*urllib*, *curl*, *wget*), but all of those get blocked.

To accomplish marketplace downloads in my work environment it would be necessary to implement an option to use this DLL for downloads, so that a connection to my company's proxy server is possible.

Do you think this can make it into VS Code? I think the two tasks would be to create a Node.js addon to make bindings for the DLL (?) and to implement the logic for downloading (could be more or less ported from the python file mentioned above) into the `download` function of [src/vs/base/node/request.ts#L79:L87](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/request.ts#L79:L87). I saw that the [`IRequestOptions`](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/request.ts#L16:L27) interface already has an `agent` member (probably for a custom user-agent).

**Comment by joaomoreno** [comment]
Thanks for this, @simonhaenisch we're definitely consider it!

**Comment by simonhaenisch** [comment]
I don't know what you changed in the meantime, but with VS Code 1.4 I can now access the extensions marketplace from within VS Code by tunneling traffic through *Fiddler* (by using it as a proxy and setting `"http.proxy": "http://127.0.0.1:8888"`). That didn't work around 6-8 weeks ago. However I still can't install extensions (*"Error: Server returned 407"*).

Let me explain to you the requests that I inspect with Fiddler. First, I do a search (e.g. for *"@popular"*) in the extensions view and VS Code makes a HTTPS request with the following content:

```http
POST https://marketplace.visualstudio.com:443/_apis/public/gallery/extensionquery HTTP/1.1
X-Market-Client-Id: VSCode 1.4.0
User-Agent: VSCode 1.4.0
X-Market-User-Id: <my id>
Content-Type: application/json
Accept: application/json;api-version=3.0-preview.1
Accept-Encoding: gzip
Content-Length: 165
Host: marketplace.visualstudio.com:443
Connection: close

{"filters":[{"criteria":[{"filterType":8,"value":"Microsoft.VisualStudio.Code"}],"pageNumber":1,"pageSize":50,"sortBy":4,"sortOrder":0}],"assetTypes":[],"flags":389}
```

*Note: I also get one HTTP `CONNECT` request for each request mentioned here, because Fiddler tunnels the requests, but I think they are irrelevant.*

All good so far, the response status is `200` and the body is as expected a json encoded list of extensions. Now I try to install an extension, e.g. *"cpptools"*, and VS Code makes a `GET` request:

```http
GET http://ms-vscode.gallery.vsassets.io:443/_apis/public/gallery/publisher/ms-vscode/extension/cpptools/0.8.1/assetbyname/Microsoft.VisualStudio.Code.Manifest HTTP/1.1
accept-encoding: gzip
Host: ms-vscode.gallery.vsassets.io:443
Connection: close
```

But this time the response status is `407` and thus another request follows, with a proxy authentication header appended:

```http
Proxy-Authorization: Negotiate TlRMT[...]==
```

The response is a `407` again (response header `X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0` for both requests). Fiddler tries to use NTLM authentication here, but the squid proxy only allows Kerberos authentication, so the proxy authentication fails.

For the first request to the marketplace, it seems that Fiddler could successfully authenticate. I don't see an initial `407` response though and there is no `Proxy-Authorization: Negotiate <Kerberos string>` header sent. I tried to request the manifest using the [url](https://ms-vscode.gallery.vsassets.io/_apis/public/gallery/publisher/ms-vscode/extension/cpptools/0.8.1/assetbyname/Microsoft.VisualStudio.Code.Manifest) with *IE11* and this way I can access the file. In that case I also don't get a `407` but a `200` response instantly. So it seems that the authentication is done some other way when using *WinINet* (which Fiddler uses, too). When I do the request from *Firefox*, I see a `407` response first, then Firefox adds a `Proxy-Authorization` header with Kerberos string and the second response is `200`, which means that Kerboros auth works. 

For comparison I did `npm install parseurl` and it also does a `GET` request pretty similar to the one to the manifest. The response status is `200`, everything works, the package gets installed. They only considerable difference is that the request also has a `accept: application/json` header, but manually adding that header to the `vsassets.io` request (by using the Fiddler request composer) didn't change anything.

I really don't understand why the request for the manifest fails to get through the proxy, while the one for the marketplace or npm works. Any help is well appreciated.

I tried setting user-agent strings in the request headers, which didn't change anything. Anyway I should get a `403` response for the second request if the user-agent wouldn't be allowed. As this seems to be unrelated to the user-agent string, should I open a new issue about this?

**Comment by joaomoreno** [comment]
Wow, great investigation! Gotta hand it to you, pretty strange proxy you've got there.

One thing I noticed in your logs was the lowercase `accept-encoding` header, for which I've pushed a fix: 5cd88ee560290daf81d20aa689f77bfbea212f5c. I wonder if it'll make a difference.

Two questions here:

- Why is it that some requests work and others don't?
- Also, could we add the `Proxy-Authorization` header, just like Firefox, to make the manifest request work? And what would that look like?

Do you have any admin in your company that could answer these?

> As this seems to be unrelated to the user-agent string, should I open a new issue about this?

It's OK, this issue already contains so much information, it would be wasteful to throw it away.

**Comment by simonhaenisch** [comment]
I think HTTP headers are case insensitive... still it's more consistent :) I changed the header manually, but couldn't request the manifest, so it's not the reason.

Unfortunately I can't copy the Kerberos `Proxy-Authorization` header from the Firefox request, because (unlike with NTML authorization) this string changes for each request afaik. I tried it and it doesn't work.

I did some more web searching and finally found a way to authenticate to the proxy using basic authentication (I didn't think the proxy allowed it, but it does). So in the end I had to base64-encode a string like `username:password` and add the following header to each request going through Fiddler:

```http
Proxy-Authorization: Basic dXNlcm5hbWU6cGFzc3dvcmQ=
```

*(The base64 string is just `username:password` encoded)*

First I got a `403` (access denied) response, so I added a permitted user-agent (Firefox) and after that I can now install extensions.

In order to append the headers to the requests I edited the `CustomRules.js` in Fiddler, adding

```js
oSession.oRequest["User-Agent"] = "<user-agent string>";
oSession.oRequest["Proxy-Authorization"] = "Basic <base64 string>";
```

to the `OnBeforeRequest` function, so they get appended to every request going through fiddler.

---

In summary these are the changes I need in VS Code (so I don't have to mess with my Fiddler settings):

* a `http.userAgent` setting to specify a custom user-agent header for requests
* a `http.proxyCredentials` setting to set a `username:password` string that gets base64-encoded and appended as proxy-authorization header with type "Basic" (or `http.username` and `http.password` settings)

There is only one problem with it, and that is that I'd need to store my password as plain text in `settings.json` and that isn't so good security-wise. Maybe you have an idea about that. I guess it would be better to not use basic auth but Kerboros instead, but unfortunately I can't make it work consistently.

**Comment by joaomoreno** [comment]
I don't think it is nice for us to change our user agent header. It's better instead to convince your network administrator to add Code as an allowed application. The proxy credentials, though, that we can do!

**Comment by VinayBanakar** [comment]
by when will this issue be fixed?
